xla/stream_executor/command_buffer.h

```diff
@@ -172,141 +105,126 @@ class CommandBuffer {
   // Command buffer API
   //===--------------------------------------------------------------------===//
 
-  // Adds an execution barrier to a given execution scope: all commands added
-  // before a barrier in a the execution scope will complete before any of the
-  // commands added after a barrier in the same execution scope.
-  virtual absl::Status Barrier(ExecutionScopeId execution_scope_id) = 0;
+  // Adds an execution barrier that depends on the commands in dep_idxes.
+  virtual absl::Status EmptyOp(Index cmd_idx, const CmdIndexSet& dep_idxes) = 0;
 
-  // Adds an execution barrier that synchronizes commands across multiple
-  // execution scopes. See example #2 in execution scope id documentation.
-  virtual absl::Status Barrier(
-      absl::Span<const ExecutionScopeId> execution_scope_ids) = 0;
+  // Adds an execution barrier that depends on all tail nodes (nodes that no
+  // other nodes depend on them) in current graph.
+  virtual absl::Status Barrier() = 0;
 
-  // Adds an execution barrier from execution scope `from_execution_scope_id` to
-  // execution scope `to_execution_scope_id`. See example #3 for details.
-  virtual absl::Status Barrier(ExecutionScopeId from_execution_scope_id,
-                               ExecutionScopeId to_execution_scope_id) = 0;
-
-  // Adds an execution barrier to the default execution scope.
-  absl::Status Barrier() { return Barrier(kDefaultExecutionScope); }
-
-  // Adds a kernel launch command.
-  virtual absl::Status Launch(ExecutionScopeId execution_scope_id,
+  // Adds a kernel launch command that depends on the commands in
+  // dep_idxes.
+  virtual absl::Status Launch(Index cmd_idx, const CmdIndexSet& dep_idxes,
                               const ThreadDim& threads, const BlockDim& blocks,
                               const Kernel& kernel, const KernelArgs& args) = 0;
 
-  // Adds a kernel launch command to the default execution scope.
-  absl::Status Launch(const ThreadDim& threads, const BlockDim& blocks,
-                      const Kernel& kernel, const KernelArgs& args) {
-    return Launch(kDefaultExecutionScope, threads, blocks, kernel, args);
-  }
+  // Adds a kernel launch command that depends on all tail nodes (nodes that no
+  // other nodes depend on them) in current graph.
+  virtual absl::Status Launch(const ThreadDim& threads, const BlockDim& blocks,
```

Comment:
It feels weird that there is no way to know what is the index of created command, I'd expect to see

absl::StatusOr<Index> Launch();

Also what's the meaning of cmd_idx above? It's kind of weird that you can accept arbitrary index from the user, I'd expect to have just one version of Launch:

absl::StatusOr<Id> Launch(const CmdDependencies& deps, ThreadDim, BlockDim, ...);

Also rename Index to Id, because it's simply a unique identifier of command within a buffer, and the fact that it's a sequential index into underlying storage is an implementation details.

Also add API like that:

// Get a set of commands that are not used as dependencies.
CmdDependencies TailDependencies();

(maybe call it LeafCommands?)